Elements (Button, Link, Paragraph)

ameseee opened the pull request:
#### What does this PR do?

- Adds rules for button, link, and paragraph elements (with variants as appropriate)
- Updates doc site to display each element_type in it's own table
- Addresses some issues brought up in [initial WIP PR](https://github.com/turingschool/savile/pull/14)

#### Where should the reviewer start?

- Screenshots
- `css/design_system/elements`

#### How should this be manually tested?

Run locally to see what is created/available, specifically on `/elements`

#### Any background context you want to provide?

It's in place and _working_ and all displaying the information needed on doc site, which is all a huge win. There are a couple things I don't love, questions I have still. I'm not sure if they should hold up the PR getting merged but wanted to note:
- The examples to explain links v buttons take up a lot of space and something about the way the page is formatted makes it ... just not super user friendly at the moment. I don't know if I used the best sub-header text to describe things. I could use some help/opinions on that!
- I wanted to make a variant for the paragraph element for dark backgrounds since that is used in many places across Turing_io - to get those to show up with a dark background, well... I didn't love my solution but it's all I could get working.
- I'd love input on the naming for the paragraph text classes. Is `s-p` too short? `s-paragraph` then those with extra tags if you will, felt super long, so I trimmed to `s-p`. 
- Also on paragraphs - notice that the sample HTML code does not display the closing HTML `<p>` tag. I can't figure out why but that's the only tag it doesn't display for.
- The multi-line code snippets are hideous; feel like that should be addressed it it's own PR or maybe grouped in with some other "style doc site" specifics.
- I'd like to make another PR that creates a minimap of sorts (screenshot from Sage below) for easier navigation within a page. As well as address our left-side bar being a little too wide IMO.
![Screen Shot 2021-03-09 at 2 18 05 PM](https://user-images.githubusercontent.com/25447342/110539371-69261600-80e2-11eb-9d7e-3b8178550a9a.png)

#### What are the relevant tickets?

- https://3.basecamp.com/3494409/buckets/19192671/todolists/3501120206
- https://3.basecamp.com/3494409/buckets/19192671/todolists/3475467357
- https://3.basecamp.com/3494409/buckets/19192671/todolists/3458194994

#### Screenshots (if appropriate)

![image](https://user-images.githubusercontent.com/25447342/110540824-30873c00-80e4-11eb-9cc9-9e7c1507facd.png)
![image](https://user-images.githubusercontent.com/25447342/110540850-38df7700-80e4-11eb-9be1-da1d27736b8a.png)
![image](https://user-images.githubusercontent.com/25447342/110540877-40068500-80e4-11eb-8185-92ddd375b799.png)

See full page screenshot here: https://www.awesomescreenshot.com/image/6513983?key=fc8eba4fa4ec214d71ed3c3710d0f00b

tannerwelsh reviewed (commented):
This is looking awesome! So cool to see it coming together.

My feedback/suggestions:

- To make the code examples render more nicely, we could use the Rouge syntax highlighter that's available to GH Pages sites: https://docs.github.com/en/github/working-with-github-pages/about-github-pages-and-jekyll#syntax-highlighting
- I'm not sure that we need to name `link-within-p` with that context-specific name. Is it bad if it is used outside of a paragraph tag? When I think of an element, unless it's something that's clearly nested (like a `<tr>`), I'd expect to be able to use it in all kinds of contexts. So I think just calling it `s-link` is enough.
- For paragraph elements, can we get away with just one paragraph element style, and use utilities for the variants? I worry that hard-coding all of the variants for elements for things like color and size will leave us with lots of repetitive element CSS, when we could get away with using utilities to combine classes to make those variants. So instead of `<p class="s-p-sm-light">`, we'd use `<p class="s-p s-font-sm s-text-light">` or something like that. Then, if we wanted to also have a small, light `<ul>` element, we wouldn't need to define `.s-ul-sm-light`, we could just use `<ul class="s-ul s-font-sm s-text-light">`.

ameseee commented:
I made updates based on 2/3 of Tanners notes above (thank you Tanner!)
- Changed link filed name to `link` and class is `s-link`
- Adjusted paragraph so there is just one base util, then devs can add any other util they want. In doc site, I showed one small example where a different size is being applied. Not sure if we'd like more examples, or context for that, maybe links to other util pages? Curious to hear your thoughts.

![image](https://user-images.githubusercontent.com/25447342/112197816-195f4880-8bd2-11eb-8c4a-2679b0063526.png)


As for the note about using rouge - yes, I plan to do that in another PR more focused on styling of the doc site itself!

ameseee commented:
@kat3kasper do you want to re-review this before it's merged?

kat3kasper commented on typography.html:
```diff
@@ -4,6 +4,7 @@
 permalink: /typography/
 ---
 
+<p>The available Typography classes only provide font-family, font-size and font-weight. See <strong>Elements -> Paragraph</strong> text for color variations on paragprah elements, and <strong>Colors</strong> for color usage recommendations with text.</p>
```
```suggestion
<p>The available Typography classes only provide font-family, font-size and font-weight. See <strong>Elements -> Paragraph</strong> text for color variations on paragraph elements, and <strong>Colors</strong> for color usage recommendations with text.</p>
```

kat3kasper replied:
Also should we link out to those pages?

kat3kasper reviewed (changes requested):
Can you update some of the screenshots on the PR so it has the latest view of those pages and can you full scroll capture them? I use Awesome Screenshot to do that

ameseee commented on _layouts/default.html:
```diff
@@ -11,7 +11,7 @@
       {% include nav.html %}
       <div class="main">
-        <h2 class="main-page-title">{{ page.title }}</h2>
+        <h2 class="s-h1">{{ page.title }}</h2>
```
Yeah, thanks for bringing this up. I think the reason we hadn't (which I'm not convinced it's the right reason by any means, but just so you know what's going through my head) is, we already have the site title "Savile Design System" in an `h1` and trying to comply with the standard that there should only 1 `h1` per page. 

I just did some reading and stand by only wanting to have one `h1` but maybe the title of the site shouldn't be the `h1`, so that the actual page title can use that tag?
- [From MDN: ](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements) Using more than one <h1> will not result in an error, but is not considered a best practice. Using only one <h1> is beneficial for screenreader users, and SEO.
- But if the site title isn't an `h1`, what tag is appropriate to use?
- Most articles and blogs I found about this are very focused on SEO. I don't imagine that's a huge priority for this but would to follow a best practice.
- I thought [a11yproject.com](https://www.a11yproject.com/resources/) would be good to compare to since I couldn't find anyone writing about this specifically... their `h1` is the page title, and name of website is inside an svg. So maybe it's ok to put "Savile Design System" in... whatever? paragraph?

kat3kasper replied:
Oh yea I don't think "Savile Design System" should be the `h1` of every page. You can put it into a `<div>` with some h1 styling if we want that. The `h1` should be the page titles. Then if we did care about SEO those would be the additional links on a google search like 'Program', 'FAQs', etc. At least in theory

<img width="663" alt="Screen Shot 2021-03-29 at 7 05 25 PM" src="https://user-images.githubusercontent.com/1551316/112918258-ba15a280-90c1-11eb-9637-c1758a281532.png">

ameseee commented on css/doc_site_styling.scss:
```diff
@@ -179,12 +179,17 @@ body {
   padding: var(--s-space-4);
 }
 
+xmp {
```
yep, I think that was noted in initial PR maybe? I'm using this as filler now before I add in rouge in another PR

ameseee commented:
This is the updated full page of the `/elements` path: 
https://www.awesomescreenshot.com/image/6921744?key=0501b625ce8a8c2f7ac88b3eae55e5a6

`/tokens` page:
https://www.awesomescreenshot.com/image/6921756?key=d12e9b6beaff9a611612886dee87b8b9

'/typography` page:
https://www.awesomescreenshot.com/image/6921779?key=c947d725870e66bcc88fbb1ec17ce13d

kat3kasper commented on _config.yml:
```diff
@@ -39,7 +39,7 @@ collections_dir: docs
 collections:
   tokens:
-    sort_by: order
+    sort_by: title
```
I don't think we want these sorted by `title` it make the different sizing of the fonts into a strange order

ameseee replied:
How do you want the whole page ordered?

kat3kasper replied:
The page should be ordered by the order the sass rules are defined:

  "ultra-light": 200,
  "light": 300,
  "normal": 400,
  "bold": 700,
  "heavy": 900
  
    "xxs": 0.625rem,
  "xs": 0.75rem,
  "sm": 0.875rem,
  "md": 1rem,
  "lg": 1.5rem,
  "xl": 2rem,
  "xxl": 3rem,

kat3kasper reviewed (changes requested):
Typically at the end of a PR I would update the PR message so that if someone was looking back at the history they wouldn't have to read through all the comments to figure out the end result of the change. 

I wouldn't trust awesome screenshot to host those screenshots forever either so next time I would download them to your computer then upload the photo so that they always exist on github.
